### ISSUE TITLE
Boot vanilla Linux distro

### ISSUE THREAD

**Issue by mdubinko** [mdubinko/gsoc-contrib#2]
Goal: identify a "vanilla" off-the-shelf Linux distro compatible with this hardware and get it to boot. Likely an image from the Einstein A20 SDK. Use this thread to identify candidates and comment on what works and doesn't work.

It would also be interesting to spin up the official GSOC app on a vanilla OS. Seeing what fails or emits error messages will give a strong signal about what's custom vs stock.

**Comment by mdubinko** [comment]
Linux sunxi : https://linux-sunxi.org/Main_Page

Installing Debian on Allwinner : https://wiki.debian.org/InstallingDebianOn/Allwinner

Cubieboard 2 Armbian : https://www.armbian.com/cubieboard-2/

HCL : Einstein A20 : https://en.opensuse.org/HCL:Einstein_A20

Android 4.2 image for Einstein A20 : https://mega.nz/folder/ZtwxCCJC#AIYHcTqz-ucjuzKnE9qD7A/folder/A9xiGQLA
Android/Jessie/Linaro images for CubieAIO : https://mega.nz/folder/ZtwxCCJC#AIYHcTqz-ucjuzKnE9qD7A/folder/ks4ShKpA
(lots of other interesting things in nearby folders ^^^)

The CubieAIO Android image is (I think) what one user on groups.io showed running on his GSOC (as shipped!).

**Comment by mdubinko** [comment]
Flashed the Android 4.2 image from the last link in previous comment to a SD. Insert and boot. It's ignored.
It may require a tool called LiveSuit to flash the image directly into the GSOC. More investigation warranted. A Linux distro may work better from SD.

**Comment by mdubinko** [comment]
debian-jessie-server-cubieaio-card-v1.0.img.7z from the same link as previous comment. Boot from SD yields black screen for ~1 minute, then finishes with text rapidly scrolling past -- only getting as far as

    Waiting for root device /dev/mmcblk1p2...

This is 3.4.something version of Linux... is it capable of reading the ext4 partition as written by version 5.4.something?

**Comment by Trinity93** [comment]
https://unix.stackexchange.com/questions/307544/boot-stuck-at-waiting-for-root-device-beagle-bone-black 

a hint... you just need to tell it your using ext4

**Comment by Trinity93** [comment]
add rootfstype=ext4 to your boot args

**Comment by RAndrewThomas** [comment]
OpenWRT might be a good choice.  I'm already familiar with it for embedded stuff plus they already have build info and images for the CubieBoard2, which is pretty close as a subset of the Einstein A20 module.  Heck Xiegu might have even used OpenWRT or at least Buildroot to develop their image.  I'll give it a shot.

Separately I added serial headers to my GSOC, one for the Einstein and one for the STM32F103.  BTW, it appears to me the STM32 is handling the hardware pushbuttons and knobs.  Probably "easy" to create new firmware for the STM32 but probably not much to gain there.

**Comment by RAndrewThomas** [comment]
Has anyone else tried connecting to the Einstein A20's serial port?  I might expect to see a serial console there and planned to try it out as soon as I got the headers on the PCB and the case back together... but I don't seem to have any TTL-USB serial converters here at work.  I'll have to try it when I get home.

**Comment by Trinity93** [comment]
BTW sorry i haven't contributed much but i currently don't have a GSOC to play with. If i can get ahold of one i'll  get stock debian running on it in no time. The allwinner A20 is soooo easy to get linux on. I just can't justify spending money on something i personally don't need atm and i am only helping out to help the community. @RAndrewThomas you are correct with the STM32 handling the knobs and buttons but they are still exposed as I/O to the linux kernel. i see no need to worry about the firmware on it and to try to hack or replace the firmware could break something unexpected.

**Comment by loredous** [comment]
> Has anyone else tried connecting to the Einstein A20's serial port? I might expect to see a serial console there and planned to try it out as soon as I got the headers on the PCB and the case back together... but I don't seem to have any TTL-USB serial converters here at work. I'll have to try it when I get home.

I have confirmed that the serial port from the board is brought out to the pins near the USB ports on the left side of the device. I am working on getting details added  to the wiki here, but in the meantime I have documented the process here: https://k0jlb.com/2021/07/23/adding-a-serial-console-to-the-gsoc/

**Comment by loredous** [comment]
> OpenWRT might be a good choice. I'm already familiar with it for embedded stuff plus they already have build info for the CubieBoard2, which is pretty close as a subset of the Einstein A20 module. Heck Xiegu might have even used OpenWRT or at least Buildroot to develop their image. I'll give it a shot.

I can confirm that Buildroot was used to develop the image, as it is referenced multiple times in various release files on the device. I am working currently on creating a buildroot configuration that will allow image creation for the GSOC. At the moment, I have been basing things off the Cubieboard2 configuration that ships with Buildroot, and have had some success in generating rootFS images. Unfortunately I have yet to be successful in building a uBoot configuration that will work on the GSOC.

I have also put in requests with CubieTech to see if we can get updated SDK information for the Einstein board specifically.

**Comment by RAndrewThomas** [comment]
> 
> 
> > Has anyone else tried connecting to the Einstein A20's serial port? I might expect to see a serial console there and planned to try it out as soon as I got the headers on the PCB and the case back together... but I don't seem to have any TTL-USB serial converters here at work. I'll have to try it when I get home.
> 
> I have confirmed that the serial port from the board is brought out to the pins near the USB ports on the left side of the device. I am working on getting details added to the wiki here, but in the meantime I have documented the process here: https://k0jlb.com/2021/07/23/adding-a-serial-console-to-the-gsoc/

Thank you.  It's nice to know it'll work when I get the box back home in a bit.  The header for the STM32 is for SWD (software debug and programming).  I have don't have any reason to go there at this time.

I'll let you know ASAP if I make relevant discoveries.

-N5RGN

**Comment by loredous** [comment]
> > > Has anyone else tried connecting to the Einstein A20's serial port? I might expect to see a serial console there and planned to try it out as soon as I got the headers on the PCB and the case back together... but I don't seem to have any TTL-USB serial converters here at work. I'll have to try it when I get home.
> > 
> > 
> > I have confirmed that the serial port from the board is brought out to the pins near the USB ports on the left side of the device. I am working on getting details added to the wiki here, but in the meantime I have documented the process here: https://k0jlb.com/2021/07/23/adding-a-serial-console-to-the-gsoc/
> 
> Thank you. It's nice to know it'll work when I get the box back home in a bit. The header for the STM32 is for SWD (software debug and programming). I have don't have any reason to go there at this time.

I'm of a similar mind on the STM32 chip for the moment. I see very little need to get into that side of things.

**Comment by RAndrewThomas** [comment]
I assume no one has figured out the root password for this device?

I've extracted "/etc/shadow" from the 1.3 firmware update image and set john to working on the root's password hash.  No other listed users have a password nor shell.

**Comment by loredous** [comment]
> I assume no one has figured out the root password for this device?

With it almost being easier to drop straight to a root shell than getting to an original login prompt, I never really felt it worth the compute cycles to break the root hash. Additionally, I would prefer to keep this project away from directly publishing what could be considered "private" information on the GSOC.

**Comment by RAndrewThomas** [comment]
Hmm, I'll keep it under my hat then.

**Comment by RAndrewThomas** [comment]
mdubinko: I'm not sure a vanilla distro gains us anything, although one for the Einstein A20 would be a good starting point.  Xiegu made their (latest version) OS using buildroot-2020.02-rc3 and there are references in the filesystem to the Cubieboard2 (though I bet there are some of those in the Einstein A20 filesystems too).  What I'm saying is it's not too hard for those with experience (I have some) to build a complete embedded distro using Buildroot.

News that might help figuring out how their app works:

- it uses the LiquidSDR library (X11/MIT license).  Home at liquidsdr.org
- it definitely uses Qt5 libraries
- they compiled their app with debug symbols!  App is /usr/local/gsoc/gsoc_app_v1

**Comment by loredous** [comment]
> mdubinko: I'm not sure a vanilla distro gains us anything, although one for the Einstein A20 would be a good starting point. Xiegu made their (latest version) OS using buildroot-2020.02-rc3 and there are references in the filesystem to the Cubieboard2 (though I bet there are some of those in the Einstein A20 filesystems too). What I'm saying is it's not too hard for those with experience (I have some) to build a complete embedded distro using Buildroot.

I agree at this point that just booting the vanilla distros for the Einstein A20 isn't going to get us much. I found similar things in my poke through the current software, in regards to Buildroot and Cubieboard2. I'd like to move this conversation over to the Discussion section, so that it doesn't get lost/hidden when this issue gets closed out. @RAndrewThomas : If you have specific experience with Buildroot, that would be fantastically helpful, as I am mostly just stumbling my way through things at this point in my work on a buildroot configuration/toolchain

**Comment by mdubinko** [comment]
@RAndrewThomas can you drop me a note at micah@dubinko.info? I'd like to ask you about something off-chain.
